### PR TITLE
[no ticket][risk=no] Puppeteer new workspace reader and writer tests

### PR DESCRIPTION
**Pull request by aweng98**
Adding new tests for READER and WRITER actions. New test covers Review Research Purpose testing.

### CODE REVIEW

File: e2e/tests/workspace/workspace-writer-actions.spec.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+    const aboutPage = await workspacesPage.openAboutPage(workspaceCard);
+    const modal = await aboutPage.openShareModal();
+    const searchInput = modal.waitForSearchBox();
+    // Verify: the Search input in Share modal is disabled.
+    expect(await searchInput.isDisabled()).toBe(true);
+    await modal.clickButton(LinkText.Cancel);
+  });
+
+  /*
```
Comment (als364) [inline e2e/tests/workspace/workspace-writer-actions.spec.ts:43]: Same comment here as above re: commented out code

File: e2e/app/page/workspaces-page.ts [hunk 2]
```diff
@@ -178,4 +181,29 @@ export default class WorkspacesPage extends AuthenticatedPage {
...
+    if (modalVisible) {
+      await reviewPurposeModal.clickReviewNowButton();
+      await aboutPage.waitForLoad();
+      // Click "Looks Good" link
+      await this.page
+        .waitForXPath('//a[text()="Looks Good"]', { visible: true, timeout: 2000 })
+        .then((link) => link.click())
+        .catch(() => {
+          // READER doesn't have permission to review research purpose. "Looks Good" link is not available to click.
+          // Ignore timeout error thrown by waitForXpath
```
Comment (als364) [inline e2e/app/page/workspaces-page.ts:201]: I guess in Puppeteer there's no other way built in to express 'doing this thing shouldn't work'?

File: e2e/tests/workspace/workspace-reader-actions.spec.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+    const aboutPage = await workspacesPage.openAboutPage(workspaceCard);
+    const modal = await aboutPage.openShareModal();
+    const searchInput = modal.waitForSearchBox();
+    // Verify: the Search input in Share modal is disabled.
+    expect(await searchInput.isDisabled()).toBe(true);
+    await modal.clickButton(LinkText.Cancel);
+  });
+
+  /*
```
Comment (als364) [inline e2e/tests/workspace/workspace-reader-actions.spec.ts:43]: I'd prefer that this commented out code either be removed or that the test just be implemented.

File: e2e/tests/workspace/workspace-reader-actions.spec.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    const homePage = new HomePage(page);
+    await homePage.getSeeAllWorkspacesLink().click();
+
+    const workspacesPage = new WorkspacesPage(page);
+    await workspacesPage.waitForLoad();
+
+    // Verify Workspace Access Level is READER.
+    const workspaceCards = await WorkspaceCard.findAllCards(page, WorkspaceAccessLevel.Reader);
+    if (workspaceCards.length === 0) {
+      return; // end test because no Workspace card available for checking
```
Comment (als364) [inline e2e/tests/workspace/workspace-reader-actions.spec.ts:25]: Why not just create a new workspace if there aren't any READER ones?

File: e2e/tests/workspace/workspace-writer-actions.spec.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    const homePage = new HomePage(page);
+    await homePage.getSeeAllWorkspacesLink().click();
+
+    const workspacesPage = new WorkspacesPage(page);
+    await workspacesPage.waitForLoad();
+
+    // Verify Workspace Access Level is WRITER.
+    const workspaceCards = await WorkspaceCard.findAllCards(page, WorkspaceAccessLevel.Writer);
+    if (workspaceCards.length === 0) {
+      return; // end test because no Workspace card available for checking
```
Comment (als364) [inline e2e/tests/workspace/workspace-writer-actions.spec.ts:25]: Same comment here as above re: no WRITER workspaces

File: e2e/app/page/workspaces-page.ts (hunk 2 above)
Comment (aweng98) [reply e2e/app/page/workspaces-page.ts:201]: No. I could write a wrapper func but I don't have many usage at this time.

File: e2e/tests/workspace/workspace-reader-actions.spec.ts (hunk 4 above)
Comment (aweng98) [reply e2e/tests/workspace/workspace-reader-actions.spec.ts:25]: I want to keep number of new workspace as low as possible. Test runs faster and leaves no footprint (not changing application data in anyway). This type of tests can be selected to run ini prod env if we want to in future.

File: e2e/tests/workspace/workspace-reader-actions.spec.ts (hunk 3 above)
Comment (aweng98) [reply e2e/tests/workspace/workspace-reader-actions.spec.ts:43]: it's placeholder to remind me without creating ticket for it. And, adding another new test makes PR larger, so I didn't want to do it now. I am okay to remove it if you still want to.

Comment (als364) [reply e2e/tests/workspace/workspace-reader-actions.spec.ts:43]: If you have plans to implement this in the near future, go ahead and leave it.

File: e2e/tests/workspace/workspace-reader-actions.spec.ts (hunk 4 above)
Comment (als364) [reply e2e/tests/workspace/workspace-reader-actions.spec.ts:25]: I guess the tester users have access to a lot of workspaces already.

**Review by als364 — APPROVED** [review]

**Comment by aweng98** [comment]
Thank you.

**Pull request merged**